src/v/cloud_storage/remote.cc

```diff
@@ -137,405 +139,607 @@ static error_outcome categorize_error(
     return result;
 }
 
-remote::remote(s3_connection_limit limit, const s3::configuration& conf)
-  : _pool(limit(), conf)
-  , _probe(remote_metrics_disabled(static_cast<bool>(conf.disable_metrics))) {}
-
-remote::remote(ss::sharded<configuration>& conf)
-  : remote(conf.local().connection_limit, conf.local().client_config) {}
-
-ss::future<> remote::start() { return ss::now(); }
-
-ss::future<> remote::stop() {
-    _as.request_abort();
-    co_await _pool.stop();
-    co_await _gate.close();
-}
+/// Retriable API method call
+///
+/// The object incapsulate all retry logic. The implementation of the api method
+/// must be derived from this template.
+/// The object is supposed to be short lived and used only for a single
+/// invocation.
+template<class Derived, class ResultT>
+class retriable {
+public:
+    using result_t = ResultT;
 
-size_t remote::concurrency() const { return _pool.max_size(); }
+    retriable(
```

Comment:
Honestly I don't like this callback-based interface, linear control flow is better. I think we can either revert to older code or try to sacrifice some e.g. log variability for simpler code (i.e. the helper can return the result, some metadata such as the number of retries to push to metrics and that's it).